jax/_src/lru_cache.py

```diff
@@ -0,0 +1,139 @@
+filelock: Any = None
+try:
+  import filelock
+except ImportError:
+  pass
+
+
+logger = logging.getLogger(__name__)
+
+
+def is_path_gs(path: str) -> bool:
+    return path.startswith("gs://")
+
+
+def get_size(path: str) -> int:
+    return tf.io.gfile.stat(path).length
+
+
+def get_mtime(path: str) -> int:
+    return tf.io.gfile.stat(path).mtime_nsec
+
+
+def update_mtime(path: str) -> None:
+    if not is_path_gs(path):
+        os.utime(path)  # set mtime (and also atime) to current time
+    else:
+        tmp_path = f"{path}.tmp"
+        tf.io.gfile.rename(path, tmp_path)
+        tf.io.gfile.rename(tmp_path, path)
+
+
+class LRUCacheImpl:
+
+    def __init__(self, path: str, *, max_size: int, timeout=10):
+        if filelock is None:
+            raise RuntimeError("Please install filelock to use the LRUCache")
+
+        self.path = path
+        self.max_size = max_size
+        os.makedirs(path, exist_ok=True)
+        self.timeout = timeout
+
+        self.lock_file = os.path.join(self.path, ".lockfile")
+        self.lock = filelock.FileLock(self.lock_file)
+
+    def get(self, key: str) -> bytes | None:
+        with self.lock.acquire(timeout=self.timeout):
+            file_path = os.path.join(self.path, key)
+
+            if not os.path.exists(file_path):
+                logger.debug("Cache miss")
+                return None
+
+            logger.debug("Cache hit")
+            update_mtime(file_path)
+            with open(file_path, "rb") as f:
+                return f.read()
+
+    def put(self, key: str, val: bytes) -> None:
+        with self.lock.acquire(timeout=self.timeout):
+            file_path = os.path.join(self.path, key)
+            if os.path.exists(file_path):
+                return
+
+            self._evict_if_needed()
+            with open(file_path, "wb") as f:
+                f.write(val)
+            update_mtime(file_path)
+
+    def _evict_if_needed(self) -> None:
+        if self.max_size == -1:
+            return  # max_size == -1: no limit on cache size
+
+        files = os.listdir(self.path)
+
+        mtime_path_sizes = []
+        dir_size = 0
+        for filename in files:
+            file_path = os.path.join(self.path, filename)
+
+            file_size = get_size(file_path)
+            file_mtime = get_mtime(file_path)
+
+            dir_size += file_size
+            mtime_path_sizes.append((file_mtime, file_path, file_size))
+
+        # sort by mtime, descending
+        mtime_path_sizes.sort(key=lambda xyz: xyz[0], reverse=True)
+
+        while dir_size >= self.max_size:
+            file_mtime, file_path, file_size = mtime_path_sizes[-1]
+            os.remove(file_path)
+            dir_size -= file_size
+            mtime_path_sizes.pop()
+
+class LRUCache(CacheInterface):
+
+  def __init__(self, path: str):
+    """
+    Sets up a cache at `path`. Cached values may already be present.
+    `path` is a directory, which could be non-empty.
+    """
+    if is_path_gs(path):
```

Comment:
I would just put this if/else logic into `get_file_cache` in compilation_cache.py, instead of making `LRUCache` a wrapper class. It's confusing right now because `LRUCache` may actually _not_ implement LRU eviction. Better to make that logic as obvious as possible to future readers of `get_file_cache`.

Reply:
Done